### DISCUSSION TITLE
Using Constraint Expressions in Firely SDK

### DISCUSSION THREAD

**Discussion by vivektk84** [FirelyTeam/firely-net-sdk#1736] [Q&A] [answered]
I am creating resources using ISourceNode and converting them to POCO. I would like to evaluate the constraint on each node and remove them before POCO.

For example, a ContactPoint datattype has this constraint:
cpt-2: A system is required if a value is provided. (expression : value.empty() or system.exists())

I want to evaluate this expression on the SourceNode that represents the contact point and get the output so that i can remove that node before creating the POCO class. (I encounter this problem where sometimes the system is not available , but the value is available when i try to map an external data to FHIR and will end up in validation failures.)

  Is there a way to achieve that using the Firely APIs?

**Comment by ewoutkramer** [comment, marked as answer]
The fhirpath evaluator you would need to evalutate the constraints needs an ITypedElement, which requires more processing than just using ISourceNode, but can be used as input to parsing it to a POCO.

So, you'd call `.ToTypedElement()` on the source node, and then run the extension method `.Select()` (Hl7.Fhir.FhirPath, namespace Hl7.Fhir.FhirPath) on it, passing it the fhirpath of the constraint.  Those methods have a small cache, so calling it multiple times with the same expression will get the compiled fhirpath statement from a cache to speed up processing.

**Reply by marcovisserFurore** [reply]
To support Ewouts answer see here some example code to run a FhirPath expression:

```c#
var json = File.ReadAllText(@"<pathToResource>/ContactPoint-example.json");

var contactPoint = FhirJsonNode.Parse(json);  // create a sourcenode
var provider = new PocoStructureDefinitionSummaryProvider(); 
var typedElement = contactPoint.ToTypedElement(provider); // provide sourcenode with type information
bool result = typedElement.Predicate("value.empty() or system.exists())"); // perform FhirPath expression
```

**Reply by vivektk84** [reply]
Thanks, @ewoutkramer @marcovisserFurore . I have tried out the above with Predicate and could get the constraints evaluated for many of them. But getting error for ele-1 where "hasValue()" is considered as invalid and also for "ext-1", the expression always returns false for an extension that has a proper format where there is a URL as well as a value[x].

Are there any restrictions on the types of constraints supported by the Predicate() method?

**Reply by ewoutkramer** [reply]
Ah, yes, let me guess, it will complain that `hasValue()` does not exist...

The default compiler for these extension methods does not know about these FHIR normative extensions to the FhirPath language.  That "just" means you need to compile the statement yourself:

``` csharp
 var symbolTable = new SymbolTable();
 symbolTable.AddStandardFP();
 symbolTable.AddFhirExtensions();

 fpCompiler = new FhirPathCompiler(symbolTable);
 var compiledExpression = compiler.Compile("the fhirpath text for the constraint here");

var instance = new ScopedNode(instanceToValidate.ToTypedElement());

var result = compiledExpression.Predicate(instance, EvaluationContext.Default());
```

Simple ;-)

**Reply by vivektk84** [reply]
Thanks @ewoutkramer .
I was trying to evaluate the below where i have passed a system , but no value eventhough its cardinality is minimum 1. The constraint evaluation for Identifer(having "ele-1") returns true. Is there any way to validate the subelements as well when you evaluate the expression for an element?That is , basically a way to evaluate the sub nodes and see if all the constraints of the subelements are also met?
![image](https://user-images.githubusercontent.com/53225232/120115370-fdaf8a00-c1a0-11eb-95e1-bb8e67c72e05.png)

**Reply by ewoutkramer** [reply]
Yes, but now you really start replicating what the full validator does. The validator will also take into account slicing (as the identifier element is sliced, and so the sets of rules to evaluate becomes conditional depending on whether this is an "UZI" identifier.

Check here for a bit of demo source code on how to run the validator: https://github.com/FirelyTeam/Furore.Fhir.ValidationDemo
